network_function/nf.c

```diff
@@ -38,85 +46,140 @@ void printPacket(char *packet, int length) {
+
+    static int i = 0;  
+    // Receive and send packets
+    while (1) {
+        char buffer[65536];
+        ssize_t length;
+
+        // Receive packet on port 1 
+        length = recvfrom(rawSocket1, buffer, sizeof(buffer), 0, NULL, NULL);
+        if (length == -1) {
+            perror("Failed to receive packet on rx port");
+            close(rawSocket1);
+            exit(1);
+        }
+        else {    
+	        i++;
+
+            // Extract the Ethernet packet type
+            unsigned short etherType = ntohs(*(unsigned short *)(buffer + 12));
+            // Print the Ethernet packet type
+            // printf("Ethernet Packet Type: 0x%04x\n", etherType);        
+
+            printf("received frame on rx port %d. Sent to tx Port\n Total NF Frame count %d\n", threadID, i);    
+
+            // Extract the source MAC address
+            unsigned char *sourceMac = buffer + 6;
+            // Print the source MAC address
+            printf("Source MAC Address: %02x:%02x:%02x:%02x:%02x:%02x\n",
+           sourceMac[0], sourceMac[1], sourceMac[2], sourceMac[3], sourceMac[4], sourceMac[5]);
+
+           //Extract the Destination MAC address
+            unsigned char *destinationMacAddress = buffer;
+            // Print the source MAC address
+            printf("Destination MAC Address: %02x:%02x:%02x:%02x:%02x:%02x\n",
+           destinationMacAddress[0], destinationMacAddress[1], destinationMacAddress[2],
+           destinationMacAddress[3], destinationMacAddress[4], destinationMacAddress[5]);     
+
+            // Send packet to port2
+            if (sendto(rawSocket2, buffer, length, 0, (struct sockaddr *)&sockAddr, sizeof(struct sockaddr_ll)) == -1) {
+                perror("Failed to send packet on port tx");
+                close(rawSocket2);
+                exit(1);
+            } 
+        }
+    }
+     pthread_exit(NULL);
+}   
+
 int main() {
-    int rawSocket;
-    struct sockaddr_ll sa_in, sa_out;
-    struct ifreq ifr_in, ifr_out;
+    int rawSocket1, rawSocket2;
+    struct sockaddr_ll sock_addr1, sock_addr2;
+    struct ifreq interface1, interface2;
+    pthread_t thread1, thread2;
+
+    struct ThreadArgs threadArgs1;
+    struct ThreadArgs threadArgs2;
+
     // Create raw socket
-    rawSocket = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
-    if (rawSocket == -1) {
-        perror("Failed to create raw socket");
+    rawSocket1 = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
+    if (rawSocket1 == -1) {
+        perror("Failed to create raw Socket_fd1");
         exit(1);
     }
-    // Set the interface to receive packets
-    strncpy(ifr_in.ifr_name, "net1", IFNAMSIZ);
-    if (ioctl(rawSocket, SIOCGIFINDEX, &ifr_in) == -1) {
+
+    rawSocket2 = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
+    if (rawSocket2 == -1) {
+        perror("Failed to create raw Socket_fd2");
+        exit(1);
+    }
+
+    // Connect to interface 1
+    strncpy(interface1.ifr_name, "net1", IFNAMSIZ);
```

Comment:
We are relying on (multus?) to rename to the two attached interfaces `net1` and `net2`. If the alias is not set and/or the rename does not happen this may break the function.

Reply:
interfaces will always be named net1, net2, net3..... and so on.
this is the standard.

Multus does provide the names, if we don't use Multus, we won't be calling any CNI's to setup the interfaces, so yes we will rely on Multus I guess.
if rename fails - it's a bug in cni-shim itself.